cli/src/etos_client/test_run/test_run.py

```diff
@@ -0,0 +1,130 @@
+        """Set up logging for ETOS remote logs."""
+        rhandler = logging.StreamHandler(sys.stdout)
+        formatter = logging.Formatter(
+            fmt="[%(rtime)s] %(levelname)s:%(rname)s: %(message)s",
+            datefmt="%Y-%m-%d %H:%M:%S",
+        )
+        rhandler.setFormatter(formatter)
+        rhandler.setLevel(loglevel)
+        self.remote_logger.propagate = False
+        self.remote_logger.addHandler(rhandler)
+
+    def __log_debug_information(self, response: ResponseSchema) -> None:
+        """Log debug information from the ETOS API response."""
+        self.logger.info("Suite ID: %s", response.tercc)
+        self.logger.info("Artifact ID: %s", response.artifact_id)
+        self.logger.info("Purl: %s", response.artifact_identity)
+        self.logger.info("Event repository: %r", response.event_repository)
+
+    def track(self, etos: ETOS, timeout: int) -> Events:
+        """Track, and wait for, an ETOS test run."""
+        end = time.time() + timeout
+
+        self.__log_debug_information(etos.response)
+        for _ in self.__log_until_eof(etos, end):
+            events = self.__collect(etos)
+            self.__status(events)
+            self.__announce(events)
+            self.__download(events)
+        self.__wait(etos, end)
+        events = self.__collect(etos)
+        self.__download(events)
+        return events
+
+    def __wait(self, etos: ETOS, timeout: int) -> None:
+        """Wait for ETOS to finish its activity."""
+        events = self.__collect(etos)
+        while not events.activity.finished:
+            self.logger.info("Waiting for ETOS to finish")
+            time.sleep(5)
+            if time.time() >= timeout:
+                raise TimeoutError("ETOS did not complete in 24 hours. Exiting")
+            events = self.__collect(etos)
+
+    def __status(self, events: Events) -> None:
+        """Check if ETOS test run has been canceled."""
+        if events.activity.canceled:
+            raise SystemExit(events.activity.canceled["data"]["reason"])
+
+    def __announce(self, events: Events) -> None:
+        """Announce current state of ETOS."""
+        self.__announcer.announce(events)
+
+    def __download(self, events: Events) -> None:
+        """Download logs and artifacts."""
+        if events.main_suites:
+            self.__downloader.download_logs(events.main_suites)
+        self.__downloader.download_artifacts(events.artifacts)
+
+    def __collect(self, etos: ETOS) -> Events:
+        """Collect events from ETOS."""
+        return self.__collector.collect(etos.response.tercc)
+
+    def __log(self, message: Message):
+        logger = getattr(self.remote_logger, message.level)
+        logger(
+            message,
+            extra={
+                "rname": message.name,
+                "rtime": message.datestring,
+            },
+        )
+
+    def __log_until_eof(self, etos: ETOS, timeout: int) -> Iterator[Ping]:
```

Comment:
It's not terribly obvious that the timeout is actually the end_time in this case. Perhaps rename the variable or document.